### PR TITLE
[Messenger] collect all stamps added on Envelope as collections

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | 4.2
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | yes
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #29156
| License       | MIT
| Doc PR        | -

Late small BC break for Messenger:
 * `Envelope::all()` takes a new optional `$stampFqcn` argument and returns the stamps for the specified FQCN, or all stamps by their class name
 * `Envelope::get()` has been renamed `Envelope::last()`

This fixes the current behavior where we replace any previous stamp with the same type, which is unexpected to me as it silently loses data and more importantly blocks interesting use cases we're going to need in the near future.
Basically, that's the same as HTTP headers being allowed to exist several times: most of them make no sense as collections, but some are really useful as collections.

### CODE REVIEW

File: src/Symfony/Component/Messenger/Envelope.php [hunk 1]
```diff
@@ -48,22 +48,26 @@ public function with(StampInterface ...$stamps): self
         $cloned = clone $this;
 
         foreach ($stamps as $stamp) {
```
Comment (ro0NL) [inline src/Symfony/Component/Messenger/Envelope.php]: while at it.. can we update with() to take at least one stamp?

Comment (nicolas-grekas) [reply src/Symfony/Component/Messenger/Envelope.php]: It works with no stamp so maybe not. Eg array_merge() is annoying to use with the coalesce operator for this reason.

Comment (ro0NL) [reply src/Symfony/Component/Messenger/Envelope.php]: hm i see, we're just make the behavior more lax with e.g. `with($somethingComputed)`

File: src/Symfony/Component/Messenger/Transport/Serialization/Serializer.php [hunk 2]
```diff
@@ -70,9 +71,8 @@ public function decode(array $encodedEnvelope): Envelope
         $stamps = $this->decodeStamps($encodedEnvelope);
 
         $context = $this->context;
-        /** @var SerializerStamp|null $serializerStamp */
-        if ($serializerStamp = $stamps[SerializerStamp::class] ?? null) {
-            $context = $serializerStamp->getContext() + $context;
+        if (isset($stamps[SerializerStamp::class])) {
+            $context = end($stamps[SerializerStamp::class])->getContext() + $context;
```
Comment (ro0NL) [inline src/Symfony/Component/Messenger/Transport/Serialization/Serializer.php]: we can actually support multiple stamps no?

File: src/Symfony/Component/Messenger/Transport/Serialization/Serializer.php [hunk 3]
```diff
@@ -34,9 +35,9 @@ class Serializer implements SerializerInterface
     private $format;
     private $context;
 
-    public function __construct(SymfonySerializerInterface $serializer, string $format = 'json', array $context = array())
+    public function __construct(SymfonySerializerInterface $serializer = null, string $format = 'json', array $context = array())
     {
-        $this->serializer = $serializer;
+        $this->serializer = $serializer ?? static::create()->serializer;
```
Comment (ro0NL) [inline src/Symfony/Component/Messenger/Transport/Serialization/Serializer.php]: weird-ish... why not inline then :) i suggest to use `Serializer::create()` in tests :thinking:

Comment (nicolas-grekas) [reply src/Symfony/Component/Messenger/Transport/Serialization/Serializer.php]: makes it easier to use, don't you think? the code is implementation detail.

File: src/Symfony/Component/Messenger/Transport/Serialization/Serializer.php (hunk 2 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Messenger/Transport/Serialization/Serializer.php]: like merging all stamps together? I don't know, so I wouldn't change that in this PR :)

Comment (ro0NL) [reply src/Symfony/Component/Messenger/Transport/Serialization/Serializer.php]: i think deciding on end() vs reset() vs merging kinda belongs to this PR

Comment (nicolas-grekas) [reply src/Symfony/Component/Messenger/Transport/Serialization/Serializer.php]: about reset() vs end(): the previous behavior returned the latest stamp, so it should be env() - in need of any reason to change it I'd keep the behavior.

File: src/Symfony/Component/Messenger/Envelope.php [hunk 4]
```diff
@@ -48,22 +48,26 @@ public function with(StampInterface ...$stamps): self
...
-            $cloned->stamps[\get_class($stamp)] = $stamp;
+            $cloned->stamps[\get_class($stamp)][] = $stamp;
         }
 
         return $cloned;
     }
 
     public function get(string $stampFqcn): ?StampInterface
```
Comment (fabpot) [inline src/Symfony/Component/Messenger/Envelope.php]: As this PR is a BC break, why not always returning an array of all stamps here?

Comment (nicolas-grekas) [reply src/Symfony/Component/Messenger/Envelope.php]: I think the use case here is very similar to HeaderBag::get() : you get only one header with, and that's nice for DX as it saves you from the extra code to deal with a collection for headers/stamps that are single.
Then, if you know a specific header/stamp is a collection, then you can get it with a different signature ->all($stampFqcn)
I think that's nice for DX so I'd suggest to keep as is.
Makes sense?

**Comment by ogizanagi** [comment]
> This fixes the current behavior where we replace any previous stamp with the same type, which is unexpected to me as it silently looses data and more importantly blocks interesting use cases we're going to need in the near future.

Nothing prevents anyone from getting and mutating/cloning a stamp with new data. So that's not blocking actually, but this suggestion is nice for DX.

Stamps were thought as an advanced feature for flexibility but them as collections is even more specific. 
DX is always nice, but does it really make sense to keep & serialize every stamps when most of them makes no sense as collections?

Couldn't stamps be mergeable by essence and a `StampInterface::merge(StampInterface $newStamp): StampInterface` used by `Envelope::with` (would either return the new stamp or a new instance with merged data, depending of the nature of the stamp)?

In case of a stamp making sense as a collection, there is no way to clear the collection (appart from re-creating an envelope). Do we care?

**Comment by nicolas-grekas** [comment]
>  does it really make sense to keep & serialize every stamps when most of them makes no sense as collections?

It does a lot to me. The comparison with HTTP headers (or with DI tags btw) is sound and meaningful.

> StampInterface::merge(StampInterface $newStamp): StampInterface

that's an interesting idea. I would do it with a separate interface that would be added to stamps that provide this method (as you said, not all of them would need that behavior). Is it worth doing when the current PR is so simple?

> no way to clear the collection (appart from re-creating an envelope).

that's legit to me, like a real-life envelope: you cannot remove a stamp. And that's on purpose: this is information about the life of the message that shouldn't be deleted - unless you put the message in a new envelope. That's a metaphor but it makes sense to me. BTW, this hints me I'd prefer the current PR vs the mergeable interface: we really don't want to lose stamps silently IMHO (when replacing)

BTW,

File: src/Symfony/Component/Messenger/Envelope.php (hunk 4 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Messenger/Envelope.php]: `Envelope::get()` has been renamed `Envelope::getLast()`

**Review by ogizanagi — APPROVED** [review]

**Review by ro0NL — APPROVED** [review]

File: src/Symfony/Component/Messenger/Envelope.php (hunk 4 above)
Comment (skalpa) [reply src/Symfony/Component/Messenger/Envelope.php]: Considering how the other methods of this class are named, shouldn't `Envelope::last()` be more appropriate ?

**Comment by skalpa** [comment]
This breaks the ability to "update" existing stamps by replacing them with a new instance. Ie:
```php
$bus->dispatch(new Envelope($msg, new SerializationStamp(array('foo' => 'bar'))));
// Later in a middleware
$context = array('bar' => 'baz');
if ($envelope->get(SerializationStamp::class)) {
    $context = array_replace($envelope->get(SerializationStamp::class)->getContext(), $context);
}
$envelope = $envelope->with(new SerializationStamp($context));
```

Also, imho, this has a lot of potential for WTF. Before the behaviour was, at least, clear: a new stamp of the same type replaces an old one. Now, stamps are added to the envelope but they may or may not be taken into account depending on whether the code that consumes them retrieves all of them or just the last one.

So, I'd push this a little further to make it more predictable:
- `Envelope::get()` should return an array
- `Envelope::getLast()` should be removed
- When several stamps of the same type are present, the explicitly defined behaviour should be that consumers should attempt to take all of them into account (by attempting to _merge_ them if that makes sense).
- Instead of adding some kind of `MergeableStampInterface::merge()`, I'd keep the individual stamps in the envelope and provide a way to merge collections at a later stage, when the stamps need to be consumed. For instance, in the case of the `SerializationStamp`, add a `public static function fromCollection(SerializationStamp ...$stamps): self`

If we compare stamps to HTTP headers, this would be in line with the HTTP behaviour that considers that duplicate headers should get merged:

> Multiple message-header fields with the same field-name MAY be present in a message if and only if the entire field-value for that header field is defined as a comma-separated list [i.e., #(values)]. It MUST be possible to combine the multiple header fields into one "field-name: field-value" pair, without changing the semantics of the message, by appending each subsequent field-value to the first, each separated by a comma.

File: src/Symfony/Component/Messenger/Envelope.php (hunk 4 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Messenger/Envelope.php]: Why not, no specific opinion on this one. I'll update if more people think last() would be better than getLast().

**Comment by skalpa** [comment]
@nicolas-grekas I understood this, but I still have to respectfully disagree. 

By only seeing the envelope as "the history of the processing of a message", you are ignoring what stamps are used for in most cases at the moment: to pass information from the outside world to the components involved in the processing of the message (ie: `SerializationStamp` and `ValidationStamp`, or in other words 100% of the stamps currently provided for user consumption by the component).

While I understand what you're trying to do and consider it positive, if from now on stamps should only be considered as a "the history of the processing of the message", then another, distinct, mean to pass configuration to middleware/senders on a message basis should be found.

I would also appreciate if you could tell me how I am supposed to fix the middleware I described above and that this PR is going to break once it gets merged, as I'm unable to find a solution by myself.

**Comment by nicolas-grekas** [comment]
I don't get the issue with the middleware example: just replace `get()` by `getLast()` and done. You shouldn't care that the previous stamp is still there: that's history nobody will care about in this case.

**Comment by skalpa** [comment]
That's right, sorry I missed that. I'm still not fully convinced by the PR implementation, but at least I'll be able to work around it, so I'll stop bothering you ;-)
Thanks for taking the time to answer me.
👍

File: src/Symfony/Component/Messenger/Envelope.php (hunk 4 above)
Comment (fabpot) [reply src/Symfony/Component/Messenger/Envelope.php]: Indeed, `last` sounds better.

Comment (nicolas-grekas) [reply src/Symfony/Component/Messenger/Envelope.php]: last() it is, updated. Thanks.

File: src/Symfony/Component/Messenger/Envelope.php [hunk 5]
```diff
@@ -48,22 +48,26 @@ public function with(StampInterface ...$stamps): self
...
-    public function get(string $stampFqcn): ?StampInterface
+    public function last(string $stampFqcn): ?StampInterface
     {
-        return $this->stamps[$stampFqcn] ?? null;
+        return isset($this->stamps[$stampFqcn]) ? end($this->stamps[$stampFqcn]) : null;
     }
 
     /**
-     * @return StampInterface[] indexed by fqcn
+     * @return StampInterface[]|StampInterface[][] The stamps for the specified FQCN, or all stamps by their class name
```
Comment (fabpot) [inline src/Symfony/Component/Messenger/Envelope.php]: Would it be better to have 2 different methods instead of two kind of returned value?

Comment (nicolas-grekas) [reply src/Symfony/Component/Messenger/Envelope.php]: not sure personally: when reading, `all()` or `all(SomeStampp:class)` is very clear to me. Discoverability looks better also (same as `EventDispatcher::getListeners()` btw) + types-wise, the interface has no method so autocompletion won't help.

Comment (skalpa) [reply src/Symfony/Component/Messenger/Envelope.php]: I think you're both right. 
Having one method is fine, but isn't grouping them by FQCN in one case a relic from how we used to manage stamps of the same type before ? 
If now stamps should also give us an history of the life of the message, it would feel more natural if `all()` (without argument) did just return a `StampInterface[]` with all the stamps in the order they were added (ie: not grouped by FQCN anymore).

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.

**Comment by sroze** [comment]
Good idea indeed 👍
